Encode the binary blob as a BigInt instead of base64

agusdallalba opened the pull request:
This is useful because QR codes encode numbers much more efficiently than base64 data.

The data URI is larger, but the QR code becomes a lot smaller!

In this case, the QR code went down from version 40 to version 30. This means you now have approximately 1217 extra bytes to play with :)

See https://huonw.github.io/blog/2024/03/qr-base10-base64/ for more information about this.

Commit:
Encode the binary blob as a BigInt instead of base64

This is useful because QR codes encode numbers much more efficiently than
base64 data.

See https://huonw.github.io/blog/2024/03/qr-base10-base64/ for more info

notpushkin commented on QRGEN.py:
```diff
@@ -10,7 +10,7 @@
 # Modified wrapper template to use gzip decompression instead of deflate
 WRAPPER_TEMPLATE = (
 """<script type="module">\
-document.open();document.write(await new Response(new Response(Uint8Array.from(atob("{b64}"), c => c.charCodeAt(0))).body.pipeThrough(new DecompressionStream("gzip"))).text());document.close();\
+document.open();document.write(await new Response(new Response(Uint8Array.fromHex({bigint}n.toString(16))).body.pipeThrough(new DecompressionStream("gzip"))).text());document.close();\
 </script>"""
```
Just to shave off a few more bytes:

```suggestion
"""<script>\
(async(R)=>\
document.write(await new R(new R(Uint8Array.fromHex({bigint}n.toString(16))).body.pipeThrough(new DecompressionStream("gzip"))).text())\
)(Response)\
</script>"""
```

notpushkin replied:
Or... we probably don’t need `document.open()` / `.close()` altogether?
Upd: yeah, seems to work! Edited suggestion.

Kuberwastaken replied:
This is weird, I can't seem to be able to play it after I put the URI data in the addresses bar

Kuberwastaken replied:
Ran about 5 different PolyThink queries, I think the main issue is that the decoded string is longer and hence the browser cuts it off or it's bigint tailing, would love if someone gets this to work though because it's seriously awesome.

notpushkin replied:
Hmm, looks like Uint8Array.fromHex is not available in Chrome yet: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array/fromHex

Is there another way to get a bigint into Uint8Array?

notpushkin replied:
```js
(async(R,n,a)=>{
  while(n){
    a.push(Number(n&255n));
    n>>=8n;
  }
  document.write(await new R(new R(new Uint8Array(a)).body.pipeThrough(new DecompressionStream("gzip"))).text())
)(Response, /* bigint */, [])
```

Edit: this one seems to be working in Chrome and Safari!

```html
data:text/html,<script>(async(R,n,a)=>{while(n){a.push(Number(n&255n));n>>=8n}document.write(await new R(new R(new Uint8Array(a.reverse())).body.pipeThrough(new DecompressionStream("gzip"))).text())})(Response,82227010980906448823283296277635751741709249650668920811527836604075455967407670585197756310497230459596445102937395493535836564385010654282046460861171301024933929737621792823910636104267801763870807302430798952182058846317586305285872438138361226954385746940530682368144147170201387945627310972279094995794321755531624609746881014821791584505830780787376019505845730822068650953053408958226475740016595439210052156274504936300896441227809713687940741210398412433987325411008011457656511967004279021817622752521490387244874182929509933726073792231595287113880418961638598042217648863389963635333826788682197266450847064453501232673597996924040680139400263789927099715005152327566323030404909628036592385509026772992889817742752766462410912405245522018573017149464594186327453134819964557985644464548947787440753561660067264906147704315235528280178979728053807165346429252381747183275624971525013336006183327647779179458648563178088651774289723701531021256290108852123834119658681762572927879273600377593361278681260178954334063417584621711105784958942143228863269381163650625380932087431874650782758341379639671364709622365761234247925497151618858240995888932854981014121604061384571022626489256088722175768010669101491637434121254911335605451918690072567349565801917180645717565080764236365437690382889189120244890028019027306602939243824050134909368415099042572690576249917899650344844370613549885844941793319997909196511521409714623993939217201614887732373884768230503327916791260447954611743473836278115764958134979001808432317447855401058356184322172357664867398574628789408055623346460839859311049133342927145575066329511107595300764906894349192426044367469669127809462597413457745663014073160418065224987619686356458431453178660116977332129858041842480652048786171335631890532968935143771425849443590038925478862676353813261198438901368050194431706953251055140363657121297891463501316147479590200689135995160587965667169512949773399663515522223050582335716248348596787996043978626704435889847809596531346756835833265875187704002288054334106853522329267807322098751188808597780927361619280212355018888090026006344505151599418979413924801109652096712696452027784220503610167572219055604739070274763930861044121775166696605778635244958781144227642631646293931195926874878745561783002239803408613015301063071118993081848353707921845493705507402857880417364211368212853787854456419541183984033595550949287934739976172570221268056941743367978198480142725392682056393311115952012398252710560150950321145777869223619984147878033668575161598175763844389098252278443959177292383738287083848095922314772920795645988855222610901949625977943499333321538268639162271770159125119763803760511683349755752731615019003557867911146064436269637504463057217195775547477532287132180609767673003974761396010344473473080823401978234677073479891254165935213786594446957878058212610164842198257964434329233130838530349151532553162427193839574358204303895374023548755184329723092370211046871340454411060768152890907433075463141006968963928019220155817443422243989580185580928480474818609927337296344107674852585157609654745754182318741073934378116146192858067187556484187370484254581077889599311869170391359980097865347362394733553632168761055655185537807849697396740587633187575146306684831714089560150968743118656637837903075547244322113335544371404155078032213221259420398170647310652042667889364793799340229649735577765549824955770932166255681178191763026885598390190080n,[])</script>
```

Kuberwastaken replied:
Oh wow! Please make the changes necessary for this and I'll merge it! This is really awesome haha

agusdallalba replied:
Ah, sorry, I didn't think of testing it in Chrome.

I'll copy notpushkin's version!

notpushkin committed:
Uint8Array.fromHex is not supported in Chrome yet

Also added whitespace to the template for easier editing

agusdallalba commented:
Pushed @notpushkin 's version that works in Chrome to my fork

Kuberwastaken commented:
This works flawlessly on Desktop but is completely unsupported on mobile.
I'm honestly not sure if this is a good change, of course, it's a huge achievement but not sure if it's worth it for less compatibility

Commit:
Percent-encode the data in the data URI.

This is probably more conservative than necessary, it percent-encodes
every character except ALPHA / DIGIT / "-" / "." / "_" / "~"

agusdallalba commented:
Hmm… It's likely we need to percent-encode the script, as it's shown in the examples at https://developer.mozilla.org/en-US/docs/Web/URI/Reference/Schemes/data

Doing this increased the QR version from 30 to 31, it's not too bad.

I've pushed another change to my fork, please try if it works.

Kuberwastaken commented:
Yeah, still no mobile compatibility, sadly

Kuberwastaken commented:
Let me put it in a different branch and mention it in the README

Kuberwastaken commented:
Added this as a branch at https://github.com/Kuberwastaken/backdooms/tree/BigInt

Kuberwastaken commented:
Closing this pull request, please feel free to add to the BigInt branch in case there's any additions :)
